projecttacoma/fqm-testify
# Add dropdown for selecting desired populations

# Summary
Allows the user to select the desired measure population(s) for each test patient.

## New Behavior
When a patient is selected, a dropdown with the label “Desired Populations” should appear.  The dropdown is populated with all the measure populations that appear in the measure bundle. When a user clicks on a measure population, it gets added to the test case state object. 

If a user clicks on a population, all of the population’s superset populations will also be selected. For example, if the numerator population is selected, then the denominator and initial population will also appear as selected values and will also get added to the test case state object.

Populations become disabled if they cannot be selected alongside the already selected populations. For example, if the user clicks on the numerator population, then the numerator exclusion population will be disabled because a patient cannot fall under both the numerator and the numerator exclusion. If the user decides to instead assign the patient to the numerator exclusion, they will need to first deselect the numerator population.

Note: Future tasking will actually use the desired populations state - this PR sets up the selection and storage in anticipation for the future tasking.

## Code Changes
* New component `utils/PopulationMultiSelect` - renders a Mantine Multiselect component. Includes functions for grabbing the measure populations off the measure bundle, keeping track of the selected populations and selecting their superset populations, and disabling populations that cannot be selected alongside already-selected populations
* Add `desiredPopulations` to the `TestCaseInfo` interface so that the desired populations for the patient to fall under can be tracked for each patient
* Add `PopulationMultiSelect` component to the `PatientInfoCard`
* Unit tests for `PatientInfoCard` and `PopulationMultiSelect`

# Testing Guidance
Run unit tests. Then, create a new test patient and test the following:
* Select the patient and select the initial population. Then, deselect the patient and select it again. Check that the initial population selection appears, showing that the desired populations state gets saved.
* Select the patient and select the denominator - check that the initial population also gets selected. This should also be the case for any other population (that when you select it, the initial population also gets selected)
    * If the denominator exclusion is available on the measure, it should be disabled.
* Select the patient and select the numerator - check that the initial population and denominator also get selected.
    * If the numerator exclusion, denominator exclusion, and/or denominator exception are available on the measure, they should be disabled.
* Select the patient and select the numerator exclusion - check that the initial population and denominator also get selected.
    * The numerator should be disabled.
* Select the patient and select the denominator exclusion - check that the initial population also gets selected
    * Everything else should be disabled.
* Select the patient and select the denominator exception - check that the initial population and denominator also get selected
    * The denominator exclusion, numerator exclusion, and numerator should be disabled. (I used the DRCommunication bundle to get access to the denominator exception for testing)

Some things to get feedback on:
* Does it make sense to disable the populations that cannot be selected with the already-selected populations? Would it be more appropriate to use the error highlighting on the multi select component instead, or maybe show a Mantine notification? Or should we get rid of the disabling altogether if we don’t want to add this checking?
* Currently, measure-population, measure-population-exclusion, and measure-observation are not included in the MultiSelect component. Should we add any handling at this time for them?
* Check that the logic for selecting/disabling measure populations aligns with the eCQM population criteria: https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html#population-criteria
